File: flutter_sdk/lib/src/data/repositories/chat_message/chat_message_repository.dart
// Copyright (c) Yalochat, Inc. All rights reserved.

import 'package:chat_flutter_sdk/src/common/page.dart';
import 'package:chat_flutter_sdk/src/common/result.dart';
import 'package:chat_flutter_sdk/src/domain/models/chat_message/chat_message.dart';

// A repository that handles chat messages in a data source.
abstract class ChatMessageRepository {
  // Method to get chat pages in descending order, a null cursor provides the first page
  // subsequent calls should include the cursor returned inside the PageInfo
  Future<Result<Page<ChatMessage>>> getChatMessagePageDesc(int? cursor, int pageSize);

  // Inserts a chat message to the data source.
  Future<Result<ChatMessage>> insertChatMessage(ChatMessage message);

  // TODO: Add clear messages
}

File: typescript-sdk/packages/core/src/data/repositories/chat-message/chat-message-repository.ts
// Copyright (c) Yalochat, Inc. All rights reserved.

import type { Result } from '../../../common/result.js';
import type { Page } from '../../../common/page.js';
import type { ChatMessage } from '../../../domain/chat-message.js';

export interface ChatMessageRepository {
  /**
   * Returns a page of messages sorted descending by id (newest first).
   * @param cursor   Last seen id — pass undefined for the first page
   * @param pageSize Number of messages per page
   */
  getChatMessagePageDesc(cursor: number | undefined, pageSize: number): Promise<Result<Page<ChatMessage>>>;

  /** Inserts a new message and returns it with the generated id. */
  insertChatMessage(message: ChatMessage): Promise<Result<ChatMessage>>;

  /** Replaces an existing message (matched by id). Returns true on success. */
  replaceChatMessage(message: ChatMessage): Promise<Result<boolean>>;
}
